Decoupled weight decay

PhilJd opened the pull request:
This PR ports the decoupled weight decay optimizers (SGDW, AdamW) to tensorflow 2.0, minus all v1 tests, as tensorflow_addons depends on tf-2.0 anyway.
Note that I factored out the testing code that is duplicated in most optimizer tests (also in base tensorflow) into `optimizer_test_base`. After this PR has been merged, I'd adapt the LazyAdam optimizer to inherit from `OptimizerTestBase`.

Sorry for the long delay!
Cheers,
Phil :)

Closes #24.

armando-fandango commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,154 @@
+
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import variables
+
```
Let's use only TF public API here if possible.

armando-fandango commented on tensorflow_addons/optimizers/weight_decay_optimizers_test.py:
```diff
@@ -0,0 +1,130 @@
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.framework import test_util
```
let's use TF public API here.

PhilJd replied:
Thanks for the comments! I initially missed that the variable initializer is available via `compat.v1.global_variables_initializer()`. I'm now using that for the graph mode tests.  I've redirected `test_util` to the `addons.test_utils`, which still uses the private API.

armando-fandango replied:
yes in `tfa.test_utils` there is a  `TODO: find public API alternative to these`

facaiy commented:
@PhilJd Welcome, thanks for the PR, and your patience. I'll take a look later this week :-)

mfojtak commented:
You probably need to add this

```
from tensorflow_addons.optimizers.weight_decay_optimizers import AdamWOptimizer
```

into *tensorflow_addons/optimizers/__init__.py* in order to expose those classes

PhilJd commented:
Thanks! I've added the respective classes and functions to `__init__.py` :)

facaiy commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,151 @@
+class OptimizerTestBase(tf.test.TestCase):
+    """Base class for optimizer tests.
+  
+  Optimizer tests may inherit from this class and define test functions
```
Could you fix doc style (align them)? `make code-format` might help you https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#development-environment

facaiy commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,151 @@
+    def doTest(self, optimizer, update_fn, params, do_sparse=False):
+        """The major test function.
+    
+    Args:
```
ditto

facaiy commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,151 @@
+                        self.evaluate(repeated_index_update_var))
+
+
+if __name__ == "__main__":
```
Do you need it?

PhilJd commented:
Thanks for the comments @facaiy, I've updated the PR!

facaiy commented on tensorflow_addons/optimizers/weight_decay_optimizers_test.py:
```diff
@@ -0,0 +1,130 @@
+
+class AdamWOptimizerTest(OptimizerTestBase):
+
+    opt_params = {
```
Can we pass them by arguments of `__init__` method? I'm fine with the current implementation, though :-)

facaiy commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,149 @@
+                    grads1 = tf.constant(grads1_np)
+                opt = optimizer(**params)
+                # Validate initial values.
+                if not tf.executing_eagerly():
```
Do we still need it?

PhilJd replied:
If we remove `test_utils.run_in_graph_and_eager_modes` and only run eager tests then we can remove the graph tests. I wasn't sure what the current best practice is and to what extent graph mode tests are still required due to `tf.function`?

facaiy replied:
I agree, perhaps we can leave a TODO comment here, remind us to remove it later. What do you think?

PhilJd replied:
I'm fine with a TODO but these tend to stay longer than often intended ;) Do you know who might be able to shed some light on this?

seanpmorgan replied:
@karmel Do you have any thoughts about this. We've been kind of all over on required tests. If we do not test with `run_in_graph_and_eager_modes` we can't support compat.v1 which is not ideal imo, but we could make that a standard.

If we do remove compat.v1 tests, should we have another util that tests everything in eager  / decorated with `tf.function`? Some of our functions are already decorated with `tf.function`, but for those that are not... there is no guarantee that a user won't add it themselves.

facaiy replied:
I change my mind after carefully reading those codes, I think they are valid because we shouldn't assume that optimizer runs inside a `tf.function` function.

facaiy commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,149 @@
+                dense update.
+        """
+        for i, dtype in enumerate([tf.half, tf.float32, tf.float64]):
+            with self.session(graph=tf.Graph()):
```
Can we remove session?

facaiy commented on tensorflow_addons/optimizers/weight_decay_optimizers.py:
```diff
@@ -0,0 +1,375 @@
+    # super().apply_x resolves to the apply_x function of the BaseOptimizer.
+
+    def _resource_apply_dense(self, grad, var):
+        with tf.control_dependencies([self._decay_weights_op(var)]):
```
Do we still need control op here? tf 2.0 will run in eager mode by default.

PhilJd commented on tensorflow_addons/optimizers/optimizer_test_base.py:
```diff
@@ -0,0 +1,149 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
```
Would you be fine with placing this class in `weight_decay_optimizers_test.py` for the period until I open the PR to refactor all tests? Then I don't need to adapt the whole code ;)

facaiy replied:
Make sense for me, Phil :-)

facaiy replied:
We forget to delete the file?

facaiy commented:
@PhilJd Phil, don't forget to run `make code-format` :-)

facaiy commented:
@PhilJd Hi, Phil. Could you address all comments? Thank you for the high quality PR, can't wait to merge it :-)

PhilJd commented:
@facaiy Thanks a lot for the comments! :)
I've addressed them all, except for the tests. I think someone put testing on the agenda for the next SIG call.
On top of that:
- I found additional documentation bugs, code samples were not compatible with tf-2.0.
- The last commit also fixes tensorflow/tensorflow#26360.
- I've renamed `AdamWOptimizer`/`SGDWOptimizer` to `AdamW`/`SGDW`, which follows the keras naming scheme instead of tf-1.0 naming scheme
- `AdamW` and `SGDW` now have a note that the decay also needs to be scheduled, originally only the extend function and the base class had this note.
- I've updated the paper title to the title of the published version

By the way, I ran `make code-format` before, but it seems this doesn't touch comments ;)
Thanks again for taking a detailed look! I also hope we can merge soon :)

facaiy commented on tensorflow_addons/optimizers/weight_decay_optimizers.py:
```diff
@@ -0,0 +1,423 @@
+      and base_optimizer.
+    """
+
+    @keras_utils.register_keras_custom_object
```
I'm afraid we have to remove the register here, because it might cause name conflict. Users have to register their custom class by themselves, and we can write the instruction in the document:
```python
Note: you might want to register your own custom optimizer by `tf.keras.utils.get_custom_objects()`.
```

facaiy commented on tensorflow_addons/optimizers/weight_decay_optimizers.py:
```diff
@@ -0,0 +1,423 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Base class to make optimizers weight decay ready."""
```
we forget to delete the file?

facaiy commented on tensorflow_addons/optimizers/weight_decay_optimizers_test.py:
```diff
@@ -0,0 +1,261 @@
+
+    def testBasicCallableParams(self):
+        self.doTest(self.optimizer, adamw_update_numpy,
+                    **self.callable_opt_params)
```
Can we rewrite the test case like what we did in SGDWTest below?

facaiy commented on tensorflow_addons/optimizers/README.md:
```diff
@@ -25,6 +27,7 @@ must:
    `@run_in_graph_and_eager_modes` (for test method)
    or `run_all_in_graph_and_eager_modes` (for TestCase subclass)
    decorator.
+ * Consider inheriting from `OptimizerTestBase`.
```
Let's remove it at first :-)

facaiy commented:
Do we forget to delete optimizer_test_base.py ?

PhilJd commented:
Oops, I forgot to commit `optimizer_test_base` ;)
Thanks for the comments!

facaiy commented:
Looks great, thank you, PhilJ! Could you resolve the merge conflict with master branch? 

By the way, you can put your name in the contact info list and code owner file if you'd like to maintain the module contributed by yourself.

facaiy commented:
@seanpmorgan @WindQAQ Sean, Tzu-Wei, do you have any concerns about this change?

seanpmorgan commented:
> @seanpmorgan @WindQAQ Sean, Tzu-Wei, do you have any concerns about this change?

No, looks like a very nice PR.. Just needs to resolve conflicts and test IMO

PhilJd commented:
Conflicts with the master are resolved and I've put my name as maintainer into the README ;)

PhilJd commented:
Interesting, running `make code-format` locally doesn't complain.

facaiy commented:
@PhilJd Phil, could you run `make code-format` again? And if you'd like, please add the line below https://github.com/tensorflow/addons/blob/4118a11f4cfebb0a3ad6449e81c5bc631d634c0d/.github/CODEOWNERS#L32

```
/tensorflow_addons/optimizers/weight_decay_optimizers*.py  @PhilJd 
```

facaiy commented:
Please ping me when you get it done, and I'll merge it. Thanks for your patience :-)

PhilJd commented:
I've applied the patch from the build server. I wasn't able to find out why the code-format doesn't complain locally. I've tried the clang format version the build task uses, `wget https://llvm.org/svn/llvm-project/cfe/trunk/tools/clang-format/git-clang-format`, but still no changes...
